### ISSUE TITLE
Intrinsically handle Expression.Property(Expression,MethodInfo)

### ISSUE THREAD

**Issue by eerhardt** [dotnet/linker#1814]
With the annotations applied in https://github.com/dotnet/runtime/pull/47938, `Expression.Property(Expression,MethodInfo)` is being marked as `RequiresUnreferencedCode`. This is because the linker doesn't see the Property being used - only the getter (or setter). See discussion in https://github.com/dotnet/runtime/pull/47938#discussion_r571314108. Thus, the linker may trim the property metadata, or the other side of the property accessor.

Because of this decision, users will now get an ILLink warning for the following code:

```C#
IQueryable<Customer> customers = GetCustomers();
foreach (var cust in customers.Where(c => c.Name == "Bob"))
{
}
```

```
C:\DotNetTest\Net6Hello\Program.cs(10,34): Trim analysis warning IL2026: Net6Hello.Program.Main(String[]): Calling 'System.Linq.Expressions.Expression.Property(Expression,MethodInfo)' which has `RequiresUnreferencedCodeAttribute` can break functionality when trimming application code. The Property metadata or other accessor may be trimmed.. [C:\DotNetTest\Net6Hello\Net6Hello.csproj]
```

The IL generated by the C# compiler to handle the above code looks like the following:

```
	𝐈𝐋_𝟎𝟎𝟑𝟑: 𝐥𝐝𝐭𝐨𝐤𝐞𝐧 𝐦𝐞𝐭𝐡𝐨𝐝 𝐢𝐧𝐬𝐭𝐚𝐧𝐜𝐞 𝐬𝐭𝐫𝐢𝐧𝐠 𝐂𝐮𝐬𝐭𝐨𝐦𝐞𝐫::𝐠𝐞𝐭_𝐍𝐚𝐦𝐞()
	𝐈𝐋_𝟎𝟎𝟑𝟖: 𝐜𝐚𝐥𝐥 𝐜𝐥𝐚𝐬𝐬 [𝐒𝐲𝐬𝐭𝐞𝐦.𝐑𝐮𝐧𝐭𝐢𝐦𝐞]𝐒𝐲𝐬𝐭𝐞𝐦.𝐑𝐞𝐟𝐥𝐞𝐜𝐭𝐢𝐨𝐧.𝐌𝐞𝐭𝐡𝐨𝐝𝐁𝐚𝐬𝐞 [𝐒𝐲𝐬𝐭𝐞𝐦.𝐑𝐮𝐧𝐭𝐢𝐦𝐞]𝐒𝐲𝐬𝐭𝐞𝐦.𝐑𝐞𝐟𝐥𝐞𝐜𝐭𝐢𝐨𝐧.𝐌𝐞𝐭𝐡𝐨𝐝𝐁𝐚𝐬𝐞::𝐆𝐞𝐭𝐌𝐞𝐭𝐡𝐨𝐝𝐅𝐫𝐨𝐦𝐇𝐚𝐧𝐝𝐥𝐞(𝐯𝐚𝐥𝐮𝐞𝐭𝐲𝐩𝐞 [𝐒𝐲𝐬𝐭𝐞𝐦.𝐑𝐮𝐧𝐭𝐢𝐦𝐞]𝐒𝐲𝐬𝐭𝐞𝐦.𝐑𝐮𝐧𝐭𝐢𝐦𝐞𝐌𝐞𝐭𝐡𝐨𝐝𝐇𝐚𝐧𝐝𝐥𝐞)
	𝐈𝐋_𝟎𝟎𝟑𝐝: 𝐜𝐚𝐬𝐭𝐜𝐥𝐚𝐬𝐬 [𝐒𝐲𝐬𝐭𝐞𝐦.𝐑𝐮𝐧𝐭𝐢𝐦𝐞]𝐒𝐲𝐬𝐭𝐞𝐦.𝐑𝐞𝐟𝐥𝐞𝐜𝐭𝐢𝐨𝐧.𝐌𝐞𝐭𝐡𝐨𝐝𝐈𝐧𝐟𝐨
	𝐈𝐋_𝟎𝟎𝟒𝟐: 𝐜𝐚𝐥𝐥 𝐜𝐥𝐚𝐬𝐬 [𝐒𝐲𝐬𝐭𝐞𝐦.𝐋𝐢𝐧𝐪.𝐄𝐱𝐩𝐫𝐞𝐬𝐬𝐢𝐨𝐧𝐬]𝐒𝐲𝐬𝐭𝐞𝐦.𝐋𝐢𝐧𝐪.𝐄𝐱𝐩𝐫𝐞𝐬𝐬𝐢𝐨𝐧𝐬.𝐌𝐞𝐦𝐛𝐞𝐫𝐄𝐱𝐩𝐫𝐞𝐬𝐬𝐢𝐨𝐧 [𝐒𝐲𝐬𝐭𝐞𝐦.𝐋𝐢𝐧𝐪.𝐄𝐱𝐩𝐫𝐞𝐬𝐬𝐢𝐨𝐧𝐬]𝐒𝐲𝐬𝐭𝐞𝐦.𝐋𝐢𝐧𝐪.𝐄𝐱𝐩𝐫𝐞𝐬𝐬𝐢𝐨𝐧𝐬.𝐄𝐱𝐩𝐫𝐞𝐬𝐬𝐢𝐨𝐧::𝐏𝐫𝐨𝐩𝐞𝐫𝐭𝐲(𝐜𝐥𝐚𝐬𝐬 [𝐒𝐲𝐬𝐭𝐞𝐦.𝐋𝐢𝐧𝐪.𝐄𝐱𝐩𝐫𝐞𝐬𝐬𝐢𝐨𝐧𝐬]𝐒𝐲𝐬𝐭𝐞𝐦.𝐋𝐢𝐧𝐪.𝐄𝐱𝐩𝐫𝐞𝐬𝐬𝐢𝐨𝐧𝐬.𝐄𝐱𝐩𝐫𝐞𝐬𝐬𝐢𝐨𝐧, 𝐜𝐥𝐚𝐬𝐬 [𝐒𝐲𝐬𝐭𝐞𝐦.𝐑𝐮𝐧𝐭𝐢𝐦𝐞]𝐒𝐲𝐬𝐭𝐞𝐦.𝐑𝐞𝐟𝐥𝐞𝐜𝐭𝐢𝐨𝐧.𝐌𝐞𝐭𝐡𝐨𝐝𝐈𝐧𝐟𝐨)
```

As you can see, the compiler grabs the metadata token for the `get_Name` getter, and then uses this to call `Expression.Property(Expression,MethodInfo)`.

We should teach the linker about this pattern so users don't get this warning.

cc @vitek-karas @MichalStrehovsky

**Comment by vitek-karas** [comment]
Currently linker will actually preserve the property metadata for such property. Basically if there's a reference to the getter (or setter) the property metadata is retained. The other accessor is fair game to trimming though. The code would probably work without the accessor available in most cases, but there's nothing in expressions which would guarantee that. The expression node stores `PropertyInfo` and no information about how the property was accessed - basically reading or writing works the same way.

So in short it's safer to add the intrinsic and mark the property as accessed via reflection - this would preserve it even if we decided to remove property metadata in other cases.

**Comment by vitek-karas** [comment]
@eerhardt 
There's probably no followup work on this on the libraries side, right? We want to keep `Expression.Property(Expression,MethodInfo)` marked with RUC since it is technically incompatible with trimming. Linker handles it with intrinsic and will avoid the warning in some cases (with #1880).

So we want to add some tests to that extent?

**Comment by eerhardt** [comment]
> We want to keep Expression.Property(Expression,MethodInfo) marked with RUC since it is technically incompatible with trimming.

That is my understanding, yes. If you had a method that took any old `MethodInfo` and passed it into `Expression.Property`, we would want to warn that the Property or "other accessor" may be trimmed. So it is unsafe by itself.

> So we want to add some tests to that extent?

What kind of tests did you have in mind? Tests that the linker warns? Doesn't warn? That your app still works after trimming?

**Comment by vitek-karas** [comment]
I think the test that we don't trim in the "Real world setup" of the Expression world. I did add a test in the linker, but those tests work very differently.

Basically something like:
```C#
var p = Expression.Property(expr, typeof(TestType).GetMethod("get_TestProp"));
((PropertyInfo)p.Member).SetValue(obj, 42);
Assert(obj.TestProp == 42);
```

**Comment by vitek-karas** [comment]
Adding the libraries trimming test in here: https://github.com/dotnet/runtime/pull/54279
